### PR TITLE
build-system: use esbuild for compiling AMP Server (new server).

### PR DESCRIPTION
**Pull request by samouri**
**summary**
- Use esbuild instead of tsc for compilation. 
- Task duration goes from **~1.3s --> ~30ms**.

Screen Shot 2021-03-12 at 5 50 36 PM

It is important to note that **esbuild does not perform typeschecking**. Therefore we still need to use `tsc`, but I've moved this usage to the `gulp check-types` command.

### CODE REVIEW

File: build-system/server/typescript-compile.js
Comment (jridgewell) [inline build-system/server/typescript-compile.js:19]: Duplicate?

File: build-system/server/typescript-compile.js
Comment (jridgewell) [inline build-system/server/typescript-compile.js:74]: `npx`?

Comment (samouri) [reply build-system/server/typescript-compile.js:74]: Does that work? I can just sub out the rest with npx? I'll try

File: build-system/server/typescript-compile.js
Comment (samouri) [reply build-system/server/typescript-compile.js:19]: Fixed

File: build-system/server/typescript-compile.js
Comment (samouri) [reply build-system/server/typescript-compile.js:74]: Of course it duz. Fixed.

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:62]: Attn: @rileyajones who is working on type-checking `build-system/` in #33246. (Perhaps at some point this step can be rolled into `check-build-system`?)

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:36]: Nice!

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:45]: If this doesn't work on Windows, this may need to be:
```suggestion
  const entryPoints = globby
    .sync(path.join(SERVER_TRANSFORM_PATH, '**', '*.ts'))
    .map(path.normalize);
```

I'm playing around with something similar for the Closure rewrite.

**Review by rsimha — COMMENTED** [review]
A few initial comments. I'd love for @rileyajones to review this and figure out whether to do the type checking here or as part of #33246.

File: build-system/server/typescript-compile.js
Comment (rileyajones) [reply build-system/server/typescript-compile.js:62]: I can definitely imagine this getting rolled into `check-build-system`. I don't think `new-server/transforms` should be checked as part of the same process that checks the reset of `build-system` as they (necessarily) use different `tsconfig.json` files but running the check in parallel makes sense.

File: build-system/server/typescript-compile.js
Comment (rileyajones) [inline build-system/server/typescript-compile.js:34]: Can this be `Promise<void>`?

Comment (samouri) [reply build-system/server/typescript-compile.js:34]: Sure. Whats the difference?

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:30]: Curious: Does this break if you don't construct the path like this? If so, I think it's more readable to use `path.normalize()` where necessary and retain the original `'build-system/server/new-server/transforms'` here.

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:66]: This message is now a part of `typecheckNewServer()`, but still talks about building. Maybe move it to `buildNewServer()` and add another failure message about type checking here?

**Review by jridgewell — APPROVED** [review]

File: build-system/tasks/check-types.js
Comment (erwinmombay) [inline build-system/tasks/check-types.js:42]: thank you!

**Review by erwinmombay — APPROVED** [review]

File: build-system/server/typescript-compile.js
Comment (rileyajones) [reply build-system/server/typescript-compile.js]: `Promise` is essentially `Promise<unknown>` which means the caller should expect something returned (that something could be an object or a string or even undefined). `Promise<void>` tells callers that the promise will resolve in nothing.

File: build-system/tasks/serve.js
Comment (rileyajones) [inline build-system/tasks/serve.js:93]: nitpick: Can you please add `@return {Promise<void>}` to these functions JSDoc?

File: build-system/tasks/server-tests.js
Comment (rileyajones) [inline build-system/tasks/server-tests.js:201]: nitpick: It would also be great if you could add a return type to this.

**Review by rileyajones — APPROVED** [review]

File: build-system/server/typescript-compile.js
Comment (samouri) [reply build-system/server/typescript-compile.js]: I modified the error to discuss type checking. I don't think I need one higher up since `esbuild` should reject.

**Comment by rileyajones** [comment]
Please wait for an LGTM from @rsimha before merging

File: build-system/server/typescript-compile.js
Comment (samouri) [reply build-system/server/typescript-compile.js]: It actually breaks because I constructed the path this way! As you informed in chat, `globby.sync` requires posix style paths.

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:43]: If `esbuild` also expects a posix path, this might work.
```suggestion
      outdir: `${SERVER_TRANSFORM_PATH}/dist`,
```

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:58]: Yay! Another `showStack` gets cleaned up.

File: build-system/server/typescript-compile.js
Comment (rsimha) [inline build-system/server/typescript-compile.js:50]: If it's functionally equivalent to call `await esbuild.build()` followed by `endBuildStep`, I think we should get rid of the `then()` chain. (We've been doing so as part of other modernization efforts across `build-system/`.)

**Review by rsimha — APPROVED** [review]
Still LGTM 😃

File: build-system/server/typescript-compile.js
Comment (rsimha) [reply build-system/server/typescript-compile.js:43]: Never mind. Looks like the current version worked nicely: https://github.com/ampproject/amphtml/pull/33247/checks?check_run_id=2125294810

**Pull request merged**
